conan/tools/files/files.py

```diff
@@ -520,3 +522,22 @@ def swap_child_folder(parent_folder, child_folder):
     child = os.path.join(parent_folder, child_folder)
     for f in os.listdir(child):
         shutil.move(os.path.join(child, f), os.path.join(parent_folder, f))
+
+
+def fix_apple_shared_install_name(conanfile):
+    def _osx_collect_dylibs(lib_folder):
+        return [f for f in os.listdir(lib_folder) if f.endswith(".dylib")
+                and not os.path.islink(os.path.join(lib_folder, f))]
+
+    def _fix_install_name(lib_name, lib_folder):
+        command = "install_name_tool -id @rpath/{} {}".format(lib_name, os.path.join(lib_folder,
```

Comment:
This is good!!

Two comments - one is about the actual install name:
We may have these 3 files:
```
libopencv_core.3.4.17.dylib
libopencv_core.3.4.dylib -> libopencv_core.3.4.17.dylib
libopencv_core.dylib -> libopencv_core.3.4.dylib
```

we need  to run the command on the file that is _not_ a symlink, but the install name is actually `/libopencv_core.3.4.dylib` - so neither the leaf name without a version, nor the one with the full version :/ What I mean is that the `lib_name` here may not coincide with the filename of the file that is not a symlink

The second one is about packages that contain multiple shared libraries, that may depend on each other. If we change the install name of a library that is depended by others, those _others_ also need to be patched to refer to the new install name. For example with opencv, for simplicity:

* `libopencv_core.dylib`
* `libopencv_imgproc.dylib`

If we inspect the properties of the second one:
```
 % otool -L libopencv_imgproc.dylib 
libopencv_imgproc.dylib:
	/path/without/patching/libopencv_imgproc.3.4.dylib (compatibility version 3.4.0, current version 3.4.17)
	/path/without/patching/libopencv_core.3.4.dylib (compatibility version 3.4.0, current version 3.4.17)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 902.1.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.100.1)
```

calling `install_name_tool -id @rpath/libopencv_core.dylib libopencv_core.dylib` will change `opencv_core` itself, but and we can also fix the install name of `opencv_imgproc`, but we still need `opencv_imgproc` to depend on `@rpath/libopencv_core.dylib` as well, which can be achieved by:

```
install_name_tool -change /path/without/patching/libopencv_core.3.4.dylib @rpath/libopencv_core.3.4.dylib
```

Otherwise we may end up with dependencies that will break at link time for consumers

Reply:
I think we can ignore the symlinks (as long as we only patch "one" file :P, so the current logic of patching the file that is not a symlink is fine.

Perhaps we could have a "logic" that "replaces" the path of the install name without affecting the leaf name?
e.g. if  `LC_ID_DYLIB` is `/path/to/library.1.2.dylib` we extract that and replace the path leading to the leaf name. I think `os.path.dirname()` can give us the bit that needs replacing -  The tricky bit is extracting the `LD_ID_DYLIB` which may require parsing the output of `otool -l` :(

Reply:
Yes, I think that's we can do 👍